x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/HashBasedTransactionStore.java

```diff
@@ -0,0 +1,491 @@
+ *      encoding (vlong)
+ * - the counts are stored as array, the index is {item_, transaction_}the id, the cell stores the count
+ * - the store can be serialized and send over the wire, serialization and deserialization uses variable length encodings to
+ *   make messages smaller
+ *
+ * ## Merging
+ *
+ * - a transaction store contains the items and transactions of 1 shard, this must be merged
+ * - transaction stores can be merged, this happens by rewriting and merging
+ *
+ * ## Pruning
+ *
+ * - prune removes all items that do not meet the support criteria and rewrites/drops transactions, this can significantly free memory
+ * - TODO: pruning is only used after merging all stores from all shards, it could be beneficial to prune already on the shard level
+ *
+ * ## Top Items/Transactions
+ *
+ * - top_{items, transactions} creates lists of {item_, transaction_}ids reverse sorted by count
+ *
+ */
+public final class HashBasedTransactionStore extends TransactionStore {
+
+    private static final Logger logger = LogManager.getLogger(HashBasedTransactionStore.class);
+
+    private static final int INITIAL_ITEM_CAPACITY = PageCacheRecycler.LONG_PAGE_SIZE;
+    private static final int INITIAL_TRANSACTION_CAPACITY = PageCacheRecycler.LONG_PAGE_SIZE;
+    private static final int CAPACITY_INCREMENT = PageCacheRecycler.LONG_PAGE_SIZE;
+
+    // re-use bytes ref object
+    private final BytesRefStreamOutput scratchItemBytesStreamOutput = new BytesRefStreamOutput();
+    private final BytesRefStreamOutput scratchTransactionBytesStreamOutput = new BytesRefStreamOutput();
+
+    // base size of transaction store, update if you add classes
+    private static final long BASE_RAM_BYTES_USED = RamUsageEstimator.shallowSizeOfInstance(HashBasedTransactionStore.class) + 2
+        * RamUsageEstimator.shallowSizeOfInstance(BytesStreamOutput.class) + 2 * RamUsageEstimator.shallowSizeOfInstance(BytesRefHash.class)
+        + 2 * RamUsageEstimator.shallowSizeOfInstance(LongArray.class);
+
+    // data holders
+    private BytesRefHash items;
+    private LongArray itemCounts;
+    private long totalItemCount;
+    private BytesRefHash transactions;
+    private LongArray transactionCounts;
+    private long totalTransactionCount;
+
+    public HashBasedTransactionStore(BigArrays bigArrays) {
+        super(bigArrays);
+        // we allocate big arrays so we have to `close` if we fail here or we'll leak them.
+        boolean success = false;
+
+        try {
+            this.items = new BytesRefHash(INITIAL_ITEM_CAPACITY, bigArrays);
+            this.itemCounts = bigArrays.newLongArray(INITIAL_ITEM_CAPACITY, true);
+            this.transactions = new BytesRefHash(INITIAL_TRANSACTION_CAPACITY, bigArrays);
+            this.transactionCounts = bigArrays.newLongArray(INITIAL_TRANSACTION_CAPACITY, true);
+            success = true;
+        } finally {
+            if (false == success) {
+                close();
```

Comment:
Does this work right?  I had the impression that if we threw from above, and then tried to enter a method instance, the object would be in an undefined state, and we might not have access to the references we need to return the memory to the breaker.  If you take a look at `HyperLogLogPlusPlus`'s constructor, we create local references for these data structures, allocate them in the try, release them explicitly in the finally, and then assign them to the members after.  I'd be more comfortable following that pattern here.

Reply:
I read the relevant section of the java spec and discussed this with @ChrisHegarty and I think we're in agreement that not using local variables here is safe as long as this class is final, which it is.  I still think this feels needlessly trappy in case we ever decide we need to extend this class, and I fail to see the harm of using locals here, but as it isn't broken right now, I'm willing to let it go.

I'm glad you have tests that randomly throw, but that does not prove that any of this is correct.  Especially since this PR has failing tests which I have not had the time to go through and verify are not related to circuit breaking.